File: legacy/artifacts/src/parsers/dex/DexMapping.ts
import { AndroidCodeUtils } from '../../android/utils/AndroidCodeUtils';

const NEWLINE = '\n'.charCodeAt(0);

export interface DexMappingClass {
  // The deobfuscated FQN of the class:
  name: string;

  // The deobfuscated signature of the class:
  signature: string;

  // The obfuscated FQN of the class:
  obfuscatedName: string;

  // The fileName if present:
  fileName?: string;

  // Start line within file. This may not be present even if fileName
  // is present.
  startLine?: number;
}

/**
 * Handles R8/Proguard mappings
 */
export class DexMapping {
  private classes: Map<string, DexMappingClass>;

  static parse(mappingBuffer: Buffer): DexMapping {
    const classes: DexMappingClass[] = [];
    let currentClass: DexMappingClass | undefined;
    const lines: string[] = [];

    let start = 0;
    for (let i = 0; i < mappingBuffer.length; ++i) {
      if (mappingBuffer[i] === NEWLINE) {
        const sub = mappingBuffer.subarray(start, i);
        const line = sub.toString('utf-8');
        lines.push(line);
        start = i + 1;
      }
    }

    // Parse the 'sourceFile' lines. The format of these lines is:
    // # {"id":"sourceFile","fileName":"CoroutineDebugging.kt"}
    const fileNameRe = /"fileName":"([^"]*)/;
    const parseComment = (line: string) => {
      // Ignore comments at the start of the file:
      if (!currentClass) {
        return;
      }
      const m = fileNameRe.exec(line);
      if (m) {
        const fileName = m[1];
        currentClass.fileName = fileName;
      }
    };

    const parseMethodOrMember = (line: string) => {
      // Format is one of:
      // originalfieldtype originalfieldname -> obfuscatedfieldname
      // [startline:endline:]originalreturntype [originalclassname.]originalmethodname(originalargumenttype,...)[:originalstartline[:originalendline]] -> obfuscatedmethodname
      // Square brackets are optional.

      // Currently we use the startLine of the first non-zero
      // method/member as the startLine of the class.
      // It's not clear if there is a more correct line number available.

      const trimmed = line.trim();
      const left = trimmed.split(' -> ')[0];
      const parts = left.split(':');
      const originalLineNumbers: number[] = [];
      for (let i = 0; i < Math.min(2, parts.length); ++i) {
        const part = parts[parts.length - i - 1];
        const n = Number(part);
        if (Number.isInteger(n)) {
          originalLineNumbers.unshift(n);
        }
      }
      const startLine = originalLineNumbers[0];
      if (currentClass) {
        if (currentClass.startLine === undefined) {
          currentClass.startLine = startLine;
        } else if (currentClass.startLine === 0) {
          currentClass.startLine = startLine;
        }
      }
    };

    const parseClass = (line: string) => {
      // Remove ':' suffix:
      line = line.slice(0, line.length - 1);

      // Split the line into obfuscated and original class names:
      const [name, obfuscatedName] = line.split(' -> ');
      if (name && obfuscatedName) {
        const signature = AndroidCodeUtils.fqnToClassSignature(name);
        const clazz = {
          name,
          signature,
          obfuscatedName,
        };

        classes.push(clazz);
        currentClass = clazz;
      }
    };

    for (const line of lines) {
      if (line.startsWith('#')) {
        parseComment(line);
      } else if (line.startsWith(' ')) {
        const trimmed = line.trim();
        if (trimmed.startsWith('#')) {
          parseComment(line);
        } else {
          parseMethodOrMember(line);
        }
      } else if (line.endsWith(':')) {
        parseClass(line);
      }
    }

    return new this(classes);
  }

  constructor(classes: DexMappingClass[]) {
    this.classes = new Map();
    for (const clazz of classes) {
      this.classes.set(clazz.obfuscatedName, clazz);
    }
  }

  size(): number {
    return this.classes.size;
  }

  deobfuscate(obfuscatedClassName: string): string | null | undefined {
    return this.classes.get(obfuscatedClassName)?.name ?? null;
  }

  deobfuscateSignature(obfuscatedSignature: string): string | null | undefined {
    return this.lookupObfuscatedSignature(obfuscatedSignature)?.signature ?? null;
  }

  lookupObfuscatedSignature(obfuscatedSignature: string): DexMappingClass | null | undefined {
    const obfuscatedFqn = AndroidCodeUtils.classSignatureToFqn(obfuscatedSignature);
    return this.lookupObfuscatedClass(obfuscatedFqn);
  }

  lookupObfuscatedClass(obfuscatedClassName: string): DexMappingClass | null | undefined {
    return this.classes.get(obfuscatedClassName);
  }

  lookupDeobfuscatedSignature(deobfuscatedClassSignature: string): DexMappingClass | null | undefined {
    const values = Array.from(this.classes.values());
    return values.find((clazz) => clazz.signature === deobfuscatedClassSignature);
  }
}

File: src/launchpad/parsers/android/dex/dex_mapping.py
import re

from dataclasses import dataclass

from launchpad.parsers.android.dex.android_code_utils import AndroidCodeUtils


@dataclass
class DexMappingClass:
    deobfuscated_fqn: str
    deobfuscated_signature: str
    obfuscated_fqn: str
    file_name: str | None = None
    # Start line within file. This may not be present even if fileName is present
    start_line: int | None = None
    # Method and field mappings for this class
    methods: dict[str, str] | None = None  # obfuscated_name -> deobfuscated_name
    fields: dict[str, str] | None = None  # obfuscated_name -> deobfuscated_name


class DexMapping:
    def __init__(self, bytes: bytes) -> None:
        self._classes: dict[str, DexMappingClass] = {}

        classes: list[DexMappingClass] = []
        current_class: DexMappingClass | None = None
        pending_file_name: str | None = None

        lines = bytes.decode("utf-8").splitlines()

        for line in lines:
            if line.startswith("#"):
                pending_file_name = self._parse_comment(line, pending_file_name)
            elif line.startswith(" "):
                trimmed = line.strip()
                if trimmed.startswith("#"):
                    pending_file_name = self._parse_comment(line, pending_file_name)
                else:
                    current_class = self._parse_method_or_member(line, current_class)
            elif line.endswith(":"):
                current_class, pending_file_name = self._parse_class(line, current_class, pending_file_name, classes)

        for clazz in classes:
            self._classes[clazz.obfuscated_fqn] = clazz

    @staticmethod
    def _parse_comment(line: str, pending_file_name: str | None) -> str | None:
        # Parse the 'sourceFile' lines. The format of these lines is:
        # # {"id":"sourceFile","fileName":"CoroutineDebugging.kt"}
        file_name_re = re.compile(r'"fileName":"([^"]*)')
        match = file_name_re.search(line)
        if match:
            return match.group(1)
        return pending_file_name

    @staticmethod
    def _parse_method_or_member(line: str, current_class: DexMappingClass | None) -> DexMappingClass | None:
        """Parse method or member lines for line number information.

        Format is one of:
        - originalfieldtype originalfieldname -> obfuscatedfieldname
        - [startline:endline:]originalreturntype [originalclassname.]originalmethodname(originalargumenttype,...)[:originalstartline[:originalendline]] -> obfuscatedmethodname
        """
        if not current_class:
            return current_class

        trimmed = line.strip()
        parts = trimmed.split(" -> ")
        if len(parts) != 2:
            return current_class

        left, obfuscated_name = parts

        # Parse line numbers
        line_parts = left.split(":")
        original_line_numbers: list[int] = []

        for i in range(min(2, len(line_parts))):
            part = line_parts[len(line_parts) - i - 1]
            try:
                n = int(part)
                original_line_numbers.insert(0, n)
            except ValueError:
                continue

        if original_line_numbers:
            start_line = original_line_numbers[0]
            if current_class.start_line is None:
                current_class.start_line = start_line
            elif current_class.start_line == 0:
                current_class.start_line = start_line

        # Parse method/field mapping
        # Remove line numbers from the left side
        if original_line_numbers:
            # Find the last colon that has a number after it
            last_colon_idx = -1
            for i, part in enumerate(line_parts):
                try:
                    int(part)
                    last_colon_idx = i
                except ValueError:
                    break

            if last_colon_idx >= 0:
                left = ":".join(line_parts[last_colon_idx + 1 :])

        # Determine if this is a method or field
        if "(" in left and ")" in left:
            if current_class.methods is None:
                current_class.methods = {}
            # Extract method name (everything before the first parenthesis)
            # Remove return type if present
            method_decl = left.split("(")[0].strip()
            if " " in method_decl:
                # e.g. 'void doSomething' -> 'doSomething'
                method_name = method_decl.split()[-1]
            else:
                method_name = method_decl
            current_class.methods[obfuscated_name] = method_name
        else:
            if current_class.fields is None:
                current_class.fields = {}
            # Extract field name (last part after the type)
            field_name = left.split()[-1].strip()
            current_class.fields[obfuscated_name] = field_name

        return current_class

    @staticmethod
    def _parse_class(
        line: str,
        current_class: DexMappingClass | None,
        pending_file_name: str | None,
        classes: list[DexMappingClass],
    ) -> tuple[DexMappingClass | None, str | None]:
        """Parse class mapping lines.

        Format: originalclassname -> obfuscatedclassname:
        """
        # Remove ':' suffix
        line = line[:-1]

        # Split the line into obfuscated and original class names
        parts = line.split(" -> ")
        if len(parts) == 2:
            name, obfuscated_fqn = parts
            if name and obfuscated_fqn:
                signature = AndroidCodeUtils.fqn_to_class_signature(name)
                clazz = DexMappingClass(
                    deobfuscated_fqn=name,
                    deobfuscated_signature=signature,
                    obfuscated_fqn=obfuscated_fqn,
                )
                # If a pending file name was found in a comment before this class, assign it
                if pending_file_name is not None:
                    clazz.file_name = pending_file_name
                    pending_file_name = None
                classes.append(clazz)
                current_class = clazz
                return current_class, pending_file_name

        return current_class, pending_file_name

    def deobfuscate(self, obfuscated_class_name: str) -> str | None:
        clazz = self._classes.get(obfuscated_class_name)
        return clazz.deobfuscated_fqn if clazz else None

    def deobfuscate_signature(self, obfuscated_signature: str) -> str | None:
        clazz = self.lookup_obfuscated_signature(obfuscated_signature)
        return clazz.deobfuscated_signature if clazz else None

    def lookup_obfuscated_signature(self, obfuscated_signature: str) -> DexMappingClass | None:
        obfuscated_fqn = AndroidCodeUtils.class_signature_to_fqn(obfuscated_signature)
        return self.lookup_obfuscated_class(obfuscated_fqn)

    def lookup_obfuscated_class(self, obfuscated_class_name: str) -> DexMappingClass | None:
        return self._classes.get(obfuscated_class_name)

    def lookup_deobfuscated_signature(self, deobfuscated_class_signature: str) -> DexMappingClass | None:
        for clazz in self._classes.values():
            if clazz.deobfuscated_signature == deobfuscated_class_signature:
                return clazz
        return None

    def deobfuscate_method(self, class_name: str, obfuscated_method_name: str) -> str | None:
        """Deobfuscate a method name for a given class.

        Args:
            class_name: The class name (can be obfuscated or deobfuscated)
            obfuscated_method_name: The obfuscated method name

        Returns:
            The deobfuscated method name, or None if not found
        """
        # First try to find the class by obfuscated name
        clazz = self.lookup_obfuscated_class(class_name)
        if clazz is None:
            # Try to find by deobfuscated signature
            clazz = self.lookup_deobfuscated_signature(class_name)
            if clazz is None:
                # Try to match by deobfuscated FQN
                for c in self._classes.values():
                    if c.deobfuscated_fqn == class_name:
                        clazz = c
                        break
                if clazz is None:
                    return None

        if clazz.methods is None:
            return None

        return clazz.methods.get(obfuscated_method_name)

    def deobfuscate_field(self, class_name: str, obfuscated_field_name: str) -> str | None:
        """Deobfuscate a field name for a given class.

        Args:
            class_name: The class name (can be obfuscated or deobfuscated)
            obfuscated_field_name: The obfuscated field name

        Returns:
            The deobfuscated field name, or None if not found
        """
        # First try to find the class by obfuscated name
        clazz = self.lookup_obfuscated_class(class_name)
        if clazz is None:
            # Try to find by deobfuscated signature
            clazz = self.lookup_deobfuscated_signature(class_name)
            if clazz is None:
                # Try to match by deobfuscated FQN
                for c in self._classes.values():
                    if c.deobfuscated_fqn == class_name:
                        clazz = c
                        break
                if clazz is None:
                    return None

        if clazz.fields is None:
            return None

        return clazz.fields.get(obfuscated_field_name)
